functional/generics-type-classes.md

```diff
@@ -0,0 +1,268 @@
+Because of the software involved, there are two different protocols you need
+to support: BOOTP (for PXE network boot), and NFS (for remote mount storage).
+
+Your goal is to have one program, written in Rust, which can handle both of
+them. It will have protocol handlers and listen for both kinds of requests. The
+main application logic will then allow a lab administrator to configure storage
+and security controls for the actual files.
+
+The requests from machines in the lab for files contain the same basic
+information, no matter what protocol they came from: an authentication method,
+and a file name to retrieve.  A straightforward implementation would look
+something like this:
+
+```rust,ignore
+struct FileDownloadRequest {
+    file_name: PathBuf,
+    authentication: Either<crate::nfs::AuthInfo, crate::bootp::AuthInfo>,
+}
+```
+
+This design might work well enough. But now suppose you needed to support
+adding metadata that was *protocol specific*. For example, with NFS, you
+wanted to determine what their mount point was in order to enforce additional
+security rules.
+
+The way the current struct is designed leaves the protocol decision until
+runtime. That means any method that applies to one protocol and not the other
+requires the programmer to do a runtime check.
+
+Here is how getting an NFS mount point would look:
+
+```rust,ignore
+struct FileDownloadRequest {
+    file_name: PathBuf,
+    authentication: Either<crate::nfs::AuthInfo, crate::bootp::AuthInfo>,
+    mount_point: Option<PathBuf>,
+}
+
+impl FileDownloadRequest {
+    // ... other methods ...
+
+    /// Gets an NFS mount point if this is an NFS request. Otherwise,
+    /// return None.
+    pub fn mount_point(&self) -> Option<&Path> {
+        self.mount_point.as_ref()
+    }
+}
+```
+
+Every caller of `get_mount point` must check for `None` and write code to handle
+it. This is true even if they know only NFS requests are ever used in a given
+code path!
+
+It would be far more optimal to cause a compile-time error if the different
+request types were confused. After all, the entire path of the user's code,
+including what functions from the library they use, will know whether a request
+is an NFS request or a BOOTP request.
+
+In Rust, this is actually possible! The solution is to *add a generic type* in
+order to split the API.
+
+Here is what that looks like:
+
+```rust
+use std::path::{Path, PathBuf};
+
+mod nfs {
+    #[derive(Clone)]
+    pub(crate) struct AuthInfo(String); // NFS session management omitted
+}
+
+mod bootp {
+    pub(crate) struct AuthInfo(); // no authentication in bootp
+}
+
+// private module, lest outside users invent their own protocol kinds!
+mod proto_trait {
+    use std::path::{Path, PathBuf};
+    use super::{bootp, nfs};
+    pub(crate) trait ProtoKind {
+        type AuthInfo;
+        fn auth_info(&self) -> Self::AuthInfo;
+    }
+    pub struct Nfs(nfs::AuthInfo, PathBuf); // the mount point metadata
```

Comment:
it's not clear that the comment is referred to the `PathBuf` element.
Maybe it's better have named struct fields?